Add support to set static IPs, gateway, DNS to wired/wireless networks

nikeflight opened the pull request:
1) As per the team discussion.

Keep the wired-network logic simple.

If eth is plugged in, enable ethernet.
If not, disable ethernet.

2) Create piaware-config.txt for debian-bookworm

a) Remove the wired-network option. No point in having it since it's based on whether an ethernet cable is physically plugged in.
b) Remove mention of dhc.

3) Added support for setting static IPs, gateway, DNS to wired/wireless networks
4) Unit Tests

nikeflight committed:
Removed logic to disable ethernet
Create piaware-config.txt for debian-bookworm

nikeflight committed:
Add support for wireless:
setting static IP
setting gateway
setting netmask
setting nameservers

eric1tran commented on scripts/piaware_config.py:
```diff
@@ -24,7 +24,7 @@
 PIAWARE_CONFIG_ENUMS["receiver"] = ["rtlsdr", "sdr", "bladerf", "beast", "relay", "radarcape", "radarcape-local", "other", "none"]
 PIAWARE_CONFIG_ENUMS["uat_receiver"] = ["sdr", "stratuxv3", "other", "none"]
-PIAWARE_CONFIG_ENUMS["network_type"] = ["static", "dhcp"]
+PIAWARE_CONFIG_ENUMS["network_type"] = ["static", "dhcp", "network-manager"]
```
Why is this needed?

nikeflight replied:
So technically, we should remove the `dhcp` option and replace it with `network-manager`. 
Then move all the default values to `network-manager`.

https://github.com/flightaware/piaware-support/pull/5/commits/afeb7b91524e9305a14dc9dff1507acb2215e776

I think this is correct since we'll be moving from dhcpd.

It's needed to differentiate between a user wanting to set static IPs or let network-manager handle all the IP assignments.

eric1tran replied:
We can discuss tomorrow, but I don't see a reason to change it to network-manager. The config options already differentiate between a user wanting a static (hardcoded) IP or a dhcp (automatically assigned) IP. 

dhcpcd is a dhcp client and network-manager has a internal dhcp client that implement the DHCP protocol.

nikeflight committed:
Added support for wired networks: setting static ip, gateway, netmask, dns
Added unit test for generate-network-config-bookworm.service

nikeflight committed:
remove dhcpd in piaware_config.py
replace default values for wireless-type and wired-type to "network-manager"

nikeflight committed:
Add start-network.service, start_network.py to up the connections after NetworkManager is started

nikeflight committed:
Merge pull request #6 from flightaware/wifi-reboot-fix

Add start-network.service, start_network.py to up the connections aft…

nikeflight committed:
Revert network-manager to dhcp in wireless/wired-type
Added functions to verify the broadcast address for wired and wireless

nikeflight committed:
Add test commands to README
Make verify_broadcast_address generic for wired and wireless

eric1tran commented on debian-bookworm/piaware-config.txt:
```diff
@@ -0,0 +1,81 @@
+# NETWORK CONFIGURATION
+#
+
+# By default, PiAware will use NetworkManager when connecting to your network.
```
No need to replace DHCP for reasons stated in my other comment. DHCP is specifying the mechanism by which the IP address is assigned.

eric1tran commented on scripts/start_network.py:
```diff
@@ -0,0 +1,21 @@
+    
+def main(dryrun=False, extra_file_path: str = None):
+    try:
+        subprocess.run(["nmcli", "con", "up", "wired"])
```
Can you see if having `autoconnect=True` under the [connection] handles what you are doing here?

nikeflight replied:
auto connect is true by default.

nikeflight replied:
But since I've moved the connection generation to before NM spins up, I may not need to manually up anything anymore.

nikeflight committed:
Add in logic to `ip addr show` then | to `grep`. Makes parsing for broadcast addresses easier.

nikeflight committed:
manually upping wireless connections is not needed. Because we create the connection profile before NM spins up
Change network-manager back to dhcp in piaware_config.py

nikeflight committed:
Change network manager back to dhcp in debian-bookworm/piaware-config.txt

nikeflight committed:
Merge branch 'wired-network-change' of https://github.com/flightaware/piaware-support into verify-brd-address

nikeflight committed:
Add an ignore_list field to MetadataSetting. Adding a file to this list will ignore any setting value for a particular key from that file.

nikeflight committed:
updated .gitignore

mutability commented on debian-bookworm/piaware-config.txt:
```diff
@@ -0,0 +1,81 @@
+
+
+# Additional settings can be added below.
+
```
Why do we have a separate default configfile for bookworm? We really want exactly one default configfile.

mutability commented on debian-bookworm/piaware-support.generate-network-config.service:
```diff
@@ -3,15 +3,15 @@
 [Unit]
 Description=Generate network configuration
-Before=network-pre.target
+Before=network-pre.target NetworkManager.service
```
Does NetworkManager.service itself have a dependency involving network-pre.target? Something here seems a bit redundant

nikeflight replied:
Some of these are holdovers from the old service.

I've removed most of them and the connections work fine.

mutability commented on debian-bookworm/piaware-support.generate-network-config.service:
```diff
@@ -3,15 +3,15 @@
 [Service]
 Type=oneshot
-ExecStart=/usr/bin/python3 /usr/lib/piaware-support/generate-network-config-bookworm.py
+ExecStart=/usr/bin/python3 /usr/lib/piaware-support/generate_network_config_bookworm.py
```
I think we want a shebang interpreter here, not explicitly refer to python3 in the service file

nikeflight replied:
As in:

Add `#!/usr/bin/python3` to generate_network_config_bookworm.py
Make it executable. (python files are not executable by default I believe?)
Call it like `/usr/lib/piaware-support/generate_network_config_bookworm.py`

nikeflight replied:
As in:

Add `#!/usr/bin/evn python3` to generate_network_config_bookworm.py.
Make it executable (chmod +x). I think we can do this in the rules folder.
Call it in this file like `ExecStart=/usr/lib/piaware-support/generate_network_config_bookworm.py`

Or is there a simpler way to do this.

eric1tran replied:
Yes I think the goal is to be able to call it in the .service file without python3 and the .py extension...i.e. `ExecStart=/usr/lib/piaware-support/generate_network_config_bookworm`

So you could just rename the file without the .py extension, add the shebang, and chmod +x it.

mutability commented on scripts/generate_network_config_bookworm.py:
```diff
@@ -0,0 +1,138 @@
+            ipv4 += configure_static_network("wired", config)
+            ipv4 += "method=manual\n"
+        except ValueError:
+            ipv4 += "method=auto\n"
```
This hides a multitude of errors in a non-obvious way. Bad config should really result in user-visible errors.

mutability commented on scripts/generate_network_config_bookworm.py:
```diff
@@ -0,0 +1,138 @@
+
+    return file
+
+def wireless_conn_file_template(config: ConfigGroup):
```
Confusing function name, this returns a populated file, not a template?

mutability commented on scripts/generate_network_config_bookworm.py:
```diff
@@ -0,0 +1,138 @@
+        ip += f"dns={format_dns(name_servers)}\n"
+    return ip
+
+def wired_conn_file_template(config: ConfigGroup):
```
Confusing function name, this returns a populated file not a template?

mutability commented on scripts/generate_network_config_bookworm.py:
```diff
@@ -0,0 +1,138 @@
+SYS_CON_DIR = "/etc/NetworkManager/system-connections"
+
+def format_dns(dns_string: str) -> str:
+    names = dns_string.strip(" \n\t")
```
This is redundant, str.split will do what you want directly:

```
>>> '  foo  \n  bar   \n'.split()
['foo', 'bar']
```

mutability commented on scripts/generate_network_config_bookworm.py:
```diff
@@ -0,0 +1,138 @@
+def configure_static_network(network_type: str, config: ConfigGroup):
+    check_address_and_netmask_set(network_type, config)
+
+    ip = ""
+    static_ip = config.get(f"{network_type}-address") + "/" + config.get(f"{network_type}-netmask")
+    gateway = config.get(f"{network_type}-gateway")
+    if gateway is not None:
+        static_ip += f",{gateway}"
+    ip += f"address1={static_ip}\n"
+
+    name_servers = config.get(f"{network_type}-nameservers")
+    if name_servers is not None:
+        ip += f"dns={format_dns(name_servers)}\n"
+    return ip
```
Since what you're doing here is building a list of lines, it might be more obvious to literally build a list of lines, then str.join() the list at the end.

mutability commented on scripts/generate_network_config_bookworm.py:
```diff
@@ -0,0 +1,138 @@
+    with open(f"{SYS_CON_DIR}/wired.nmconnection", "w") as conn_file:
+        conn_file.write(wired_conn_file_template(config))
+
+    subprocess.run(["chmod", "600", f"{SYS_CON_DIR}/wired.nmconnection"])
```
use `os.chmod`

mutability commented on scripts/generate_network_config_bookworm.py:
```diff
@@ -0,0 +1,138 @@
+        conn_file.write(wired_conn_file_template(config))
+
+    subprocess.run(["chmod", "600", f"{SYS_CON_DIR}/wired.nmconnection"])
+    subprocess.run(["sync"])
```
If we're going to sync, let's do it once at the very end.

Also, use `os.sync`

mutability commented on scripts/generate_network_config_bookworm.py:
```diff
@@ -0,0 +1,138 @@
+    with open(f"{SYS_CON_DIR}/wireless.nmconnection", "w") as conn_file:
+        conn_file.write(wireless_conn_file_template(config))
+
+    subprocess.run(["chmod", "600", f"{SYS_CON_DIR}/wireless.nmconnection"])
+    subprocess.run(["sync"])
```
same comments as above

mutability commented on scripts/generate_network_config_bookworm.py:
```diff
@@ -0,0 +1,138 @@
+    return file
+
+def generate_wired_network_config(config: ConfigGroup):
+    with open(f"{SYS_CON_DIR}/wired.nmconnection", "w") as conn_file:
```
Why is this not conditional on `wired-network yes`?

mutability commented on scripts/generate_network_config_bookworm.py:
```diff
@@ -0,0 +1,138 @@
+    if not config.get("wireless-network"):
+        print("wireless-network set to no")
+        return
+
```
If you change from `wireless-network yes` to `wireless-network no`, then without a cleanup step, this will leave behind the old `wireless.nmconnection` with the old config.

I mentioned this in slack earlier and I thought you said the files were written unconditionally? If not then you need cleanup.

nikeflight replied:
Yes I got my wires crossed.

We can write the files unconditionally. The autoconnect field will determine if the wireless/wired network will be enabled/disabled.

mutability commented on scripts/piaware_config.py:
```diff
@@ -57,20 +60,21 @@ def __init__(self, **kwargs):
+        self.settings["wired-address"] = MetadataSettings(sdonly=True, network=True, setting_type="str")
+        self.settings["wired-netmask"] = MetadataSettings(sdonly=True, network=True, setting_type="str")
+        # Setting broadcast address directly through boot/firmare/piaware-config.txt has been deprecated.
+        self.settings["wired-broadcast"] = MetadataSettings(sdonly=True, network=True, setting_type="str", ignore_list=[BOOT_PIAWARE_CONF])
```
Why is this setting ignored, but only from specific files?

nikeflight replied:
I added this as a way to deprecate or ignore settings.

I guess it makes sense to just have the logic be `ignore any settings for this value from all files`.

nikeflight replied:
I've changed this to just be a `deprecated` field.

I don't see a situation where we want to deprecate a setting in one file but not another.

mutability commented on scripts/piaware_config.py:
```diff
@@ -57,20 +60,21 @@ def __init__(self, **kwargs):
-        self.settings["wireless-gateway"] = MetadataSettings(sdonly=True, network=True)
-        self.settings["wireless-nameservers"] = MetadataSettings(default = ["8.8.8.8", "8.8.4.4"], sdonly=True, network=True)
+        self.settings["wireless-address"] = MetadataSettings(sdonly=True, network=True, setting_type="str")
+        self.settings["wireless-broadcast"] = MetadataSettings(sdonly=True, network=True, setting_type="str", ignore_list=[BOOT_PIAWARE_CONF])
```
as above

mutability commented on scripts/piaware_config.py:
```diff
@@ -57,20 +60,21 @@ def __init__(self, **kwargs):
-        self.settings["wired-broadcast"] = MetadataSettings(sdonly=True, network=True)
-        self.settings["wired-gateway"] = MetadataSettings(sdonly=True, network=True)
-        self.settings["wired-nameservers"] = MetadataSettings(default= ["8.8.8.8", "8.8.4.4"], sdonly=True, network=True)
+        self.settings["wired-address"] = MetadataSettings(sdonly=True, network=True, setting_type="str")
```
A couple of wider comments about this `__init__`:

1) use a dict literal, rather than laboriously initializing each key
2) why do these dicts need to be unique per instance? Can you initialize this just once, at the class leve, and share the result?

nikeflight replied:
Updated. This was a straight carry over from the old tcl config.

mutability commented on scripts/piaware_config.py:
```diff
@@ -210,6 +214,13 @@ def validate_uuid(self, val: str, version=4) -> bool:
     def validate_gain(self, val: str) -> bool:
         return self.validate_double(val) or (isinstance(val, str) and val == "max")
 
+    def check_if_file_in_ignore_list(self, key, file_name) -> bool:
```
If you defaulted `ignore_list` to `[]` then this function becomes just `return (file_name in self.settings[key].ignore_list)` which is brief enough you may not even need a function for it.

mutability reviewed (changes requested):
Some initial feedback. I need to do a wider review of the older changes at some point as I haven't looked at those in detail yet.

nikeflight committed:
Addressing PR comments
Moved MetadataSettings to be a dict literal
Change subprocess calls to use os module
We write unconditionally to connection files. We set autoconnect variable in order to enable/disable the network
Remove debian-bookworm/piaware-config.txt

nikeflight committed:
PR comments
Fix unit tests
Fix dict initialization

nikeflight committed:
Merge branch 'wired-network-change' of https://github.com/flightaware/piaware-support into verify-brd-address

nikeflight committed:
PR comments
Verify brd address against what the user set

nikeflight committed:
Add shebang to generate_network_config_bookworm.py
Call it directly in systemd service

nikeflight committed:
Merge pull request #7 from flightaware/verify-brd-address

Verify brd address

nikeflight committed:
PR comments

eric1tran commented on debian-bookworm/piaware-support.generate-network-config.service:
```diff
@@ -3,15 +3,11 @@
 [Service]
 Type=oneshot
-ExecStart=/usr/bin/python3 /usr/lib/piaware-support/generate-network-config-bookworm.py
+ExecStart=/usr/lib/piaware-support/generate_network_config_bookworm.py
```
You can drop `.py` extension from generate_network_config_bookworm.py file, the .service file, and the .install file

nikeflight committed:
drop .py extension from generate_network_config_bookworm

nikeflight committed:
Change ignore_list to deprecated field

nikeflight committed:
Clean up dict init

nikeflight committed:
Chnage netmask to be dotted decimal format

eric1tran reviewed (commented):
Did some blackbox testing below and seem to be working as expected.
- Switched to a different wireless-ssid
- Toggling wired-network with Ethernet cable connected/disconnected
- Toggling wireless-network
- Static IP configuration for wireless and wired

The only difference I noticed from before is that `wireless-netmask` went from optional to required if you want to use static IP. Is this needed?

nikeflight commented:
> Did some blackbox testing below and seem to be working as expected.
> 
> * Switched to a different wireless-ssid
> * Toggling wired-network with Ethernet cable connected/disconnected
> * Toggling wireless-network
> * Static IP configuration for wireless and wired
> 
> The only difference I noticed from before is that `wireless-netmask` went from optional to required if you want to use static IP. Is this needed?

I believe people should specify there net masks when setting a static IP.

That being said, we can also assign a default net mask to them.
Network Manager seems to assign a default net mask of 32 (not super helpful).

I believe class A IP4 addresses typically have a net mask of 8.
B = 16
C = 24.

dhcpcd follows this convention. https://github.com/NetworkConfiguration/dhcpcd/issues/164

eric1tran commented:
Ah, thanks for the reference! So dhcpcd infers it based on the IP address. 

I think it's reasonable to do the same by providing a default. We can probably go with a class c netmask 255.255.255.0 (/24) given most users are probably on small home networks. I think it would provide a better user experience than setting just the wireless-address and then finding out the network wasn't configured because it now needs wireless-netmask.

mutability commented:
plenty of home networks use 10.x.x.x/8, the default if not provided should definitely follow the class-based rules at a minimum.

nikeflight committed:
Add in default netmask for class A,B,C static IPs

nikeflight committed:
Clean up unused settings
Add support for adept-serverhosts

nikeflight committed:
Revert "Clean up unused settings"

This reverts commit a27c1a05fbfa6338ac3e98de42879bd77d9209b9.

nikeflight committed:
We actually need the settings that I removed.
Just set adept-serverhosts setting_type to str

nikeflight committed:
Add in WHITEOUT logic

nikeflight committed:
Add netmask validation
Add update unit tests for netmask -> Decimal Dot notation

nikeflight committed:
PR comments

eric1tran committed:
Update piaware-restart-network-bookworm script and generalize name when installing it

eric1tran committed:
Fix typo

nikeflight committed:
Moved validation and parsing to class types

nikeflight committed:
Add method to read config

nikeflight committed:
Move piaware_config.py to own directory
Move generate_network_config_bookworm, .py file to their own directory

eric1tran committed:
Merge pull request #8 from flightaware/update_piaware_network_restart_script

Update piaware-network-restart-bookworm script

nikeflight committed:
Fixed get_prefix
Other PR comments

eric1tran commented on flightaware_piaware_config/pyproject.toml:
```diff
@@ -0,0 +1,3 @@
+[project]
+name = "flighaware_piaware_config"
```
spelling

nikeflight committed:
spelling

nikeflight committed:
Fixed parsing line with comment

nikeflight committed:
piaware-restart-network-bookworm: Add logic to up/down connections so changes can be applied
